flush/sink.go

```diff
@@ -0,0 +1,176 @@
+package flush
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/uw-labs/substrate"
+	"github.com/uw-labs/substrate-tools/message"
+	"github.com/uw-labs/sync/rungroup"
+)
+
+const (
+	defaultMsgBufferSize = 50
+	defaultAckBufferSize = 100
+)
+
+// An AsyncMessageSinkOption is a function which sets an AsyncMessageSink configuration option.
+type AsyncMessageSinkOption func(ams *AsyncMessageSink)
+
+// WithAckFunc set the AckFunc callback function, which is called for each ack recieved. If the
+// AckFunc returns an error, it will be returned by `Run`.
+func WithAckFunc(ackFn AckFunc) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.ackFn = ackFn
+	}
+}
+
+// WithMsgBufferSize sets the msg channel buffer size.
+func WithMsgBufferSize(size int) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.msgBufferSize = size
+	}
+}
+
+// WithAckBufferSize sets the ack channel buffer size.
+func WithAckBufferSize(size int) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.ackBufferSize = size
+	}
+}
+
+// AckFunc is a callback function executed by `Run` for each message succesfully produced. The ack
+// counter is only incremented once this function returns. If no AckFunc is provided the ack counter
+// is always incremented.
+type AckFunc func(msg substrate.Message) error
+
+// AsyncMessageSink wraps substrate.AsyncMessageSink and provides an interface for interaction with
+// the underlying sink, as well as the capability to flush the message buffer.
+type AsyncMessageSink struct {
+	ctx           context.Context
+	cancel        context.CancelFunc
+	wait          func() error
+	sink          substrate.AsyncMessageSink
+	msgs          uint64
+	msgBufferSize int
+	msgCh         chan substrate.Message
+	acks          uint64
+	ackBufferSize int
+	ackCh         chan substrate.Message
+	ackFn         AckFunc
+}
+
+// NewAsyncMessageSink returns a pointer a new AsyncMessageSink.
+// See examples/flush/sink.go for example usage.
+func NewAsyncMessageSink(ctx context.Context, sink substrate.AsyncMessageSink, opts ...AsyncMessageSinkOption) *AsyncMessageSink {
+	ctx, cancel := context.WithCancel(ctx)
+
+	ams := &AsyncMessageSink{
+		ctx:           ctx,
+		cancel:        cancel,
+		sink:          sink,
+		msgBufferSize: defaultMsgBufferSize,
+		ackBufferSize: defaultAckBufferSize,
+	}
+
+	for _, opt := range opts {
+		opt(ams)
+	}
+
+	ams.msgCh = make(chan substrate.Message, ams.msgBufferSize)
+	ams.ackCh = make(chan substrate.Message, ams.ackBufferSize)
+
+	return ams
+}
+
+// Run initialises message publishing using the underlying sink and blocks until either an error
+// occurs or the constructor context is done. If an AckFunc is configured, Run will execute it for
+// each ack recieved. If an error is returned, the user should cancel the constructor context to
+// prevent `Flush` from blocking.
+func (ams *AsyncMessageSink) Run() error {
```

Comment:
Instead of exposing `Run`, why not run automatically on construction (in another goroutine) and exit that goroutine on `Close` (and possibly also on error)

Reply:
It complicates error handling a bit, for example which error do we return on close if sarama gives an error on close, or other goroutine produces an error?

It also gives users an option to choose when to start the engine. For example we have a thing which starts producing / consuming on GRPC API call, so we can do `Run()` on API call, etc

Reply:
I agree it would be slightly more complex to implement, including the error handling you mention. I'm wondering though whether it would overall be simpler from the user perspective.

It also avoids thinking about:

Reply:
The problem is that we need to register consumer only when API is called not on the startup, because everything is dynamic. We could make it part of the constructor, but blocking on constructor seems wrong, so this would be a spin off on its own goroutine and then more issues come to life.

On example what if user doesn't call `Close()` and sarama craps out? close should return errors for closing part not the running part of the code.

The way we do this right now is:

```
sink:= flush.NewAsyncSink()
go func(){
   err := sink.Run()
   if err != nil{
       log.Panic(err)
   }
}
```

and doing this seems wrong:

```
sink:= flush.NewAsyncSink()
go func(){
   err := sink.Close()
}
```

Reply:
The way sarama got around this problem is `sink.Errors() <-chan *ProducerError` method on sink, we could do that, but it adds a bit more complexity and keeps the same behaviour:

```
sink:= flush.NewAsyncSink()
go func(){
   err := <- sink.Errors()
   if err != nil{
       log.Panic(err)
   }
}
```

Reply:
Additionally the substrate behaviour is similiar to our implementation, first you call a constructor then you run `PublishMessages(ctx context.Context, acks chan<- Message, messages <-chan Message) error` which in our case named `Run()` instead of `PublishMessages`. So substrate users are already familiar with the pattern

Ref: https://github.com/uw-labs/substrate/blob/master/substrate.go#L37